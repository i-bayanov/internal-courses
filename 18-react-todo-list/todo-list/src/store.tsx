import { configureStore } from '@reduxjs/toolkit';
import { IAction } from './interfaces-and-types';
import todoReducer from './reducer';

const store = configureStore({ reducer: todoReducer });

const dispatchAndSave = (action: IAction) => {
  store.dispatch(action);
  localStorage.setItem('todos', JSON.stringify(store.getState()));
};

type RootState = ReturnType<typeof store.getState>;

export { store, dispatchAndSave, RootState };
