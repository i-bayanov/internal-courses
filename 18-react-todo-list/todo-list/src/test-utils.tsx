import React from 'react';
import { render } from 'react-dom';
import { BrowserRouter } from 'react-router-dom';

import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';

import todoReducer from './reducer';

const STORE = [
  {
    id: '1651234007958',
    title: 'Write a to-do list',
    completed: true,
  },
  {
    id: '1651234009086',
    title: 'Mark the first item as completed',
    completed: true,
  },
  {
    id: '1651234010110',
    title: '-',
    completed: false,
  },
  {
    id: '1651234864772',
    title: 'Come up with an interesting third point',
    completed: false,
  },
];

const mockStore = configureStore({ reducer: todoReducer, preloadedState: STORE });

// eslint-disable-next-line no-undef
const testRender = (component: JSX.Element, container: Element) => render(
  <Provider store={mockStore}>
    <BrowserRouter>
      {component}
    </BrowserRouter>
  </Provider>,
  container,
);

export { STORE, testRender };
