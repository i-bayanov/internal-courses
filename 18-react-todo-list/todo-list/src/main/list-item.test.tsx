import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';

import ListItem from './list-item';
import { STORE } from '../test-utils';
import * as store from '../store';

let container: Element | null = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container!);
  container!.remove();
  container = null;
});

describe('List item component', () => {
  it('should renders correctly', () => {
    render(<ListItem {...STORE[0]} />, container!);
    expect(container).toMatchSnapshot();
  });

  it('List item component should change on double click', () => {
    const spy = jest.spyOn(store, 'dispatchAndSave');

    render(<ListItem {...STORE[0]} />, container);
    container?.firstElementChild?.dispatchEvent(new MouseEvent('dblclick', { bubbles: true }));
    const input = container?.querySelector('.edit') as HTMLInputElement;
    input.value = 'Write a to-do list parapam';
    input.dispatchEvent(new KeyboardEvent('keydown', {
      key: 'Enter',
      bubbles: true,
    }));
    expect(spy).toHaveBeenCalledWith({
      type: 'todo/edit',
      payload: {
        id: '1651234007958',
        value: 'Write a to-do list parapam',
      },
    });
    jest.restoreAllMocks();
  });
});
