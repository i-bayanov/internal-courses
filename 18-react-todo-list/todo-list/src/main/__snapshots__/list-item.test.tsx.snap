// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List item component should renders correctly 1`] = `
<div>
  <li
    class="completed "
    data-id="1651234007958"
  >
    <div
      class="view"
    >
      <input
        checked=""
        class="toggle"
        type="checkbox"
      />
      <label>
        Write a to-do list
      </label>
      <button
        class="destroy"
      />
    </div>
  </li>
</div>
`;
